### ISSUE TITLE
`hist_approx_count_below` includes values > threshold 

### ISSUE THREAD

**Issue by pradeepcrao** [openhistogram/libcircllhist#72]
The description for `hist_approx_count_below` reads ` Returns the number of values in buckets that are entirely lower than or equal to threshold`. However, currently the method includes values  greater than threshold.

Consider the values `109, 110, 111` stored in the histogram.  The counts for the buckets will be as follows:
```

                         bucket.val   bucket.exp  count
hist->bvs[0]:                    10           2       1
hist->bvs[1]:                    11           2       2

```

For a threshold of 110, [this line](https://github.com/openhistogram/libcircllhist/blob/master/src/circllhist.c#L699) will return true for both buckets, even though clearly the value 111 stored in the second bucket is greater than the threshold.

That line should read `if (threshold < bucket_lower)` to ensure values over the threshold are not included in the count.

**Comment by bdunavant** [comment]
I believe this is a documentation error.   The above PR should fix the docs and add clarify that `count_above(x) + count_below(x) == count()`

**Comment by bdunavant** [comment]
I will note that there is no "right" answer here, because of the lost resolution in a histogram.  A more complete answer would be new/modified functions that allow the caller to specify if they want to include, or exclude buckets that don't match perfectly on a boundary, or perhaps include a portion of the bucket, based on an assumed/specified distribution of samples in the bucket.   That's out of scope of my PR though, which clarifies the intent of the original code.   I'll leave this up for comment for a few days, and then merge.

**Comment by pradeepcrao** [comment]
Unfortunately, we were relying on the old behaviour to get the count strictly [below](https://github.com/envoyproxy/envoy/blob/main/envoy/stats/histogram.h#L70) quantile buckets [here.](https://github.com/envoyproxy/envoy/blob/main/source/common/stats/histogram_impl.cc#L97)

What do you feel about me adding a method that gives us the old behaviour named say `hist_approx_count_strictly_below` that effectively is a copy of `hist_approx_count_below` with the comparison changed to `if (threshold < bucket_lower)` ?

**Comment by jmarantz** [comment]
Note about above: we noticed this because upgrading this library broke tests for Envoy (https://github.com/envoyproxy/envoy). Pradeep says he can work around by making two calls but this is actually going to impact proxy performance, and we'd prefer not to make our histogram processing slower.

**Comment by bdunavant** [comment]
Pradeepcrao's suggestion of a new function that works the way he wants (and is documented as to the difference) seems right to me, but maybe slightly different proposal:
Add:  
`hist_approx_count_below_inclusive`
`hist_approx_count_below_exclusive`
`hist_approx_count_above_inclusive`
`hist_approx_count_above_exclusive`
Change:
`count_above` to be an alias to `hist_approx_count_above_exclusive`
`count_below` to be an alias to `hist_approx_count_below_inclusive`

I think this should give you what you need for your use-case, and the complementary one, as well as add clarity in code.   Would that work for you?  Pradeepcrao would you like to implement that?

**Comment by bdunavant** [comment]
I'm seeing your post that it changed, and I do see a change in the behavior 4 years ago in a commit.   Lemme research that a bit more.

**Comment by bdunavant** [comment]
Theo has made changes in the above PR.

**Comment by pradeepcrao** [comment]
Thank you, `hist_approx_count_below_exclusive` is what we need.

**Comment by postwait** [comment]
Closed with #74
